rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * SonarLint Core - RPC Implementation
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.rpc.impl;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.lsp4j.jsonrpc.CompletableFutures;
+import org.springframework.beans.factory.BeanFactory;
+
+class AbstractRpcServiceDelegate {
+
+  private final Supplier<BeanFactory> beanFactorySupplier;
+  private final ExecutorService requestsExecutor;
+  private final ExecutorService requestAndNotificationsSequentialExecutor;
+
+  protected AbstractRpcServiceDelegate(Supplier<BeanFactory> beanFactorySupplier, ExecutorService requestsExecutor, ExecutorService requestAndNotificationsSequentialExecutor) {
+    this.beanFactorySupplier = beanFactorySupplier;
+    this.requestsExecutor = requestsExecutor;
+    this.requestAndNotificationsSequentialExecutor = requestAndNotificationsSequentialExecutor;
+  }
+
+  protected <T> T getBean(Class<T> clazz) {
+    return beanFactorySupplier.get().getBean(clazz);
+  }
+
+  protected <R> CompletableFuture<R> requestAsync(Function<CancelChecker, R> code) {
+    return CompletableFutures.computeAsync(requestAndNotificationsSequentialExecutor, cancelChecker -> {
+      var wrapper = new CancelCheckerWrapper(cancelChecker);
+      wrapper.checkCanceled();
+      return wrapper;
+    }).thenApplyAsync(cancelChecker -> {
+      cancelChecker.checkCanceled();
+      return code.apply(cancelChecker);
+    }, requestsExecutor);
+  }
+
+  protected <R> CompletableFuture<R> runAsync(Consumer<CancelChecker> code) {
```

Comment:
We could lose the generic type and just use `CompletableFuture<Void>` 
(The callers also have `CompletableFuture<Void>` as a return type)

Reply:
Good catch, that was the point of this method :)